articles/budougumi0617.re

```diff
@@ -1,15 +1,450 @@
+もっと平易な言葉で言い直すと、@<tt>{クラス（型）の階層構造（親子関係）による継承}です。
+代表的なオブジェクト指向言語である@<tt>{Java}でサブクラシングの例を書いたコードが@<list>{person}です。
+@<list>{person}は、親となる@<code>{Person}クラスと子となる@<code>{Japanese}クラスの定義と、@<code>{Person}クラスを引数に取るメソッドを含んでいます。
+
+//list[person][Javaで表現されたPersonクラスを継承するJapaneseクラス][c#]{
+class Person {
+  String name;
+  int age;
+}
+
+// Personクラスを継承したJapaneseクラス
+class Japanese extends Person {
+  int myNumber;
+}
+
+class Main {
+  // Personクラスを引数にとるメソッド
+  public static void Hello(Person p) {
+    System.out.println("Hello " + p.name);
+  }
+}
+//}
+
+@<code>{Person}クラスを継承した@<code>{Japanese}クラスのオブジェクトは、ポリモフィズムによって@<code>{Person}変数に代入できます。
+また、同様に@<code>{Person}クラスのオブジェクトを引数にとるメソッドに対して代入することもできます（@<list>{person2}）。
+
+//list[person2][Javaにおけるクラス継承を利用したポリモフィズムな代入と呼び出し]{
+Japanese japanese = new Japanese();
+person.name = "budougumi0617";
+Person person = japanese;
+Main.Hello(japanese);
+//}
+
+このような継承関係（ポリモフィズム）を表現するとき、Goはインターフェースを使うでしょう。
+Goは@<list>{person}のような実像クラス（あるいは抽象クラス）を親とするようなサブクラシングによる継承の仕組みを言語仕様としてサポートしていません。
+Goでは埋込み（@<i>{Embedding}@<fn>{embedding}）を使って別の型に実装を埋め込むアプローチもあります。
+しかし、これは多態性や共変性・反変性@<fn>{convariance}を満たしません。
+よって、埋め込みはオブジェクト指向で期待される継承ではなくコンポジションにすぎません。
+@<list>{go_person}は@<list>{person}をGoで書き直したものです。
+@<list>{go_person}の例では、@<code>{Japanese}型のオブジェクトは@<code>{Hello}関数に利用することはできません。
+
+//list[go_person][Goで@<list>{person}のような親子関係を表現する場合]{
+type Person struct {
+  Name string
+  Age  int
+}
+
+// Personを埋め込んだJapanese型。
+type Japanese struct {
+  Person
+  MyNumber int
+}
+
+func Hello(p Person) {
+  fmt.Println("Hello " + p.Name)
+}
+//}
+以上の例以外にも、@<kw>{リスコフの置換原則}などの一部のSOLIDの原則をそのままGoに適用することはできません。
+しかし、SOLIDの原則のベースとなる考えを取り入れることでよりシンプルで可用性の高いGoのコードを書くことは可能です。
+
+それでは、次節よりGoのコードにSOLIDの原則を適用していくとどうなっていくか見ていきます。
+
+
+//footnote[convariance][@<href>{https://docs.microsoft.com/ja-jp/dotnet/csharp/programming-guide/concepts/covariance-contravariance/}]
+//footnote[embedding][@<href>{https://golang.org/doc/effective_go.html#embedding}]
+
+===[column] 実装よりもコンポジションを選ぶ
+あるクラスが他の具象クラス（抽象クラス）を拡張した場合の継承を、@<tt>{Java}の世界では@<kw>{実装継承}（@<i>{implemebtation inheritance}）と呼びます。
+あるクラスがインターフェースを実装した場合や、インターフェースが他のインターフェースを拡張した場合の継承を@<kw>{インターフェース継承}（@<i>{interface inheritance}）と呼びます。
+Goがサポートしている継承はJavaの言い方を借りるならば@<kw>{インターフェース継承}のみです。
+クラスの親子関係による@<kw>{実装継承}はカプセル化を破壊する危険も大きく、深い継承構造はクラスの構成把握を困難にするという欠点もあります。
+このことは代表的なオブジェクト指向言語である@<kw>{Java}の名著、「@<kw>{Effective Java}」の「@<kw>{項目16 継承よりコンポジションを選ぶ}」でも言及されています（筆者が所有している@<kw>{Effective Java}は第2版ですが、2018年に@<tt>{Java 9}対応の<kw>{Effective Java 第3版}が発売されています）。
+Goが@<kw>{実装継承}をサポートしなかった理由は明らかになっていませんが、筆者は以上の危険性があるためサポートされていないと考えています。
+
+===[/column]
+
+== @<kw>{単一責任の原則}（@<kw>{SRP}, @<tti>{Single responsibility principle}）
+//quote{
+クラスを変更する理由は1つ以上存在してはならない。
+//}
+
+ここでは@<tt>{アジャイルソフトウェア開発の奥義}の原文どおりに「クラスを…」とそのまま引用しましたが、
+@<kw>{単一責任の原則}はさまざまな粒度で考えなければいけない問題です。
+関数・メソッド単位ならば比較的考えやすそうです。パッケージや型（クラス）の単位になってくると
+
+
+変更理由が1つしか存在しないということは、そのクラスの役割（責任）がひとつであることを意味します。
+では、「このクラスの役割が1つである」と判断するにはどうすればよいのでしょうか？
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Disallow to use "？". (ja-technical-writing/no-exclamation-question-mark)